### PR TITLE
Amend function calls that use fr_pair_t ** and fr_pair_t * for lists to use fr_pair_list_t *

### PR DESCRIPTION
**Pull request by ndptech**

### CODE REVIEW

File: src/lib/eap_aka_sim/state_machine.c [hunk 1]
```diff
@@ -3320,7 +3320,7 @@ static rlm_rcode_t common_reauthentication_response_recv_resume(module_ctx_t con
  *
  * This is called by the state_* functions to decode the peer's response.
  */
-static rlm_rcode_t common_decode(fr_pair_t **subtype_vp, fr_pair_t **vps,
+static rlm_rcode_t common_decode(fr_pair_list_t *subtype_vp, fr_pair_list_t *vps,
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:3323]: subtype_vp is returning a vp it's not a fr_pair_list_t

File: src/lib/ldap/base.h [hunk 2]
```diff
@@ -468,7 +468,7 @@ char const	*fr_ldap_edir_errstr(int code);
 /*
  *	map.c - Attribute mapping code.
  */
-int		fr_ldap_map_getvalue(TALLOC_CTX *ctx, fr_pair_t **out, request_t *request,
+int		fr_ldap_map_getvalue(TALLOC_CTX *ctx, fr_pair_list_t *out, request_t *request,
```
Comment (arr2036) [inline src/lib/ldap/base.h:471]: returning a vp, not a pair list

File: src/lib/ldap/map.c [hunk 3]
```diff
@@ -36,7 +36,7 @@ USES_APPLE_DEPRECATED_API
  *
  * @see map_to_vp
  */
-int fr_ldap_map_getvalue(TALLOC_CTX *ctx, fr_pair_t **out, request_t *request, map_t const *map, void *uctx)
+int fr_ldap_map_getvalue(TALLOC_CTX *ctx, fr_pair_list_t *out, request_t *request, map_t const *map, void *uctx)
```
Comment (arr2036) [inline src/lib/ldap/map.c:39]: returning a vp, not a pair list

File: src/lib/server/map.c [hunk 4]
```diff
@@ -828,7 +828,7 @@ void map_sort(map_t **maps, fr_cmp_t cmp)
  *	- 0 on success.
  *	- -1 on failure.
  */
-static int map_exec_to_vp(TALLOC_CTX *ctx, fr_pair_t **out, request_t *request, map_t const *map)
+static int map_exec_to_vp(TALLOC_CTX *ctx, fr_pair_list_t *out, request_t *request, map_t const *map)
```
Comment (arr2036) [inline src/lib/server/map.c:831]: returning a vp, not a list

File: src/lib/server/map.c [hunk 5]
```diff
@@ -908,7 +908,7 @@ static int map_exec_to_vp(TALLOC_CTX *ctx, fr_pair_t **out, request_t *request,
  *	- 0 on success.
  *	- -1 on failure.
  */
-int map_to_vp(TALLOC_CTX *ctx, fr_pair_t **out, request_t *request, map_t const *map, UNUSED void *uctx)
+int map_to_vp(TALLOC_CTX *ctx, fr_pair_list_t *out, request_t *request, map_t const *map, UNUSED void *uctx)
```
Comment (arr2036) [inline src/lib/server/map.c:911]: returning a vp, not a list

File: src/lib/tls/ocsp.c [hunk 6]
```diff
@@ -133,7 +133,7 @@ static int ocsp_staple_from_pair(request_t *request, SSL *ssl, fr_pair_t *vp)
  *	- -1 on error.
  *	- 0 on success.
  */
-static int ocsp_staple_to_pair(fr_pair_t **out, request_t *request, OCSP_RESPONSE *resp)
+static int ocsp_staple_to_pair(fr_pair_list_t *out, request_t *request, OCSP_RESPONSE *resp)
```
Comment (arr2036) [inline src/lib/tls/ocsp.c:136]: Returns a single pair, not a pair list

File: src/lib/unlang/xlat_builtin.c [hunk 7]
```diff
@@ -75,7 +75,7 @@ static fr_sbuff_parse_rules_t const xlat_arg_parse_rules = {
  *	- -4 if either the attribute or qualifier were invalid.
  *	- The same error codes as #tmpl_find_vp for other error conditions.
  */
-int xlat_fmt_get_vp(fr_pair_t **out, request_t *request, char const *name)
+int xlat_fmt_get_vp(fr_pair_list_t *out, request_t *request, char const *name)
```
Comment (arr2036) [inline src/lib/unlang/xlat_builtin.c:78]: Emits a single vp not a pair list

File: src/lib/unlang/xlat_builtin.c [hunk 8]
```diff
@@ -108,7 +108,7 @@ int xlat_fmt_get_vp(fr_pair_t **out, request_t *request, char const *name)
  *	- -4 if either the attribute or qualifier were invalid.
  *	- The same error codes as #tmpl_find_vp for other error conditions.
  */
-int xlat_fmt_copy_vp(TALLOC_CTX *ctx, fr_pair_t **out, request_t *request, char const *name)
+int xlat_fmt_copy_vp(TALLOC_CTX *ctx, fr_pair_list_t *out, request_t *request, char const *name)
```
Comment (arr2036) [inline src/lib/unlang/xlat_builtin.c:111]: Emits a single vp, not a pair list

File: src/modules/rlm_lua/lua.c [hunk 9]
```diff
@@ -194,7 +194,7 @@ DIAG_ON(type-limits)
  *	- 0 on success.
  *	- -1 on failure.
  */
-static int fr_lua_unmarshall(fr_pair_t **out,
+static int fr_lua_unmarshall(fr_pair_list_t *out,
```
Comment (arr2036) [inline src/modules/rlm_lua/lua.c:197]: Emits a single VP, not a pair list

File: src/lib/eap_aka_sim/state_machine.c [hunk 10]
```diff
@@ -880,7 +880,7 @@ static int common_encode(request_t *request, eap_session_t *eap_session, uint16_
 	eap_session->this_round->request->id = eap_aka_sim_session->id++ & 0xff;
 	eap_session->this_round->set_request_id = true;
 
-	ret = fr_aka_sim_encode(request, head, &encoder_ctx);
+	ret = fr_aka_sim_encode(request, &head, &encoder_ctx);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:883]: Extra level of indirection

File: src/lib/eap_aka_sim/state_machine.c [hunk 11]
```diff
@@ -3499,7 +3499,7 @@ static rlm_rcode_t common_reauthentication(module_ctx_t const *mctx, request_t *
 	 *	Case 1 where we're allowed to send an EAP-Failure
 	 */
 	case FR_SUBTYPE_VALUE_AKA_SIM_CLIENT_ERROR:
-		client_error_debug(request, from_peer);
+		client_error_debug(request, &from_peer);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:3502]: Extra level of indirection

File: src/lib/eap_aka_sim/state_machine.c [hunk 12]
```diff
@@ -3623,7 +3623,7 @@ static rlm_rcode_t aka_challenge(module_ctx_t const *mctx, request_t *request)
...
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:3626]: Extra level of indirection

File: src/lib/eap_aka_sim/state_machine.c [hunk 13]
```diff
@@ -3683,7 +3683,7 @@ static rlm_rcode_t sim_challenge(module_ctx_t const *mctx, request_t *request)
...
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:3686]: Extra level of indirection

File: src/lib/eap_aka_sim/state_machine.c [hunk 14]
```diff
@@ -3795,7 +3795,7 @@ static rlm_rcode_t aka_identity(module_ctx_t const *mctx, request_t *request)
 	 *	identity.
 	 */
 	case FR_SUBTYPE_VALUE_AKA_SIM_CLIENT_ERROR:
-		client_error_debug(request, from_peer);
+		client_error_debug(request, &from_peer);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:3798]: Extra level of indirection

File: src/lib/eap_aka_sim/state_machine.c [hunk 15]
```diff
@@ -3900,7 +3900,7 @@ static rlm_rcode_t sim_start(module_ctx_t const *mctx, request_t *request)
...
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:3903]: Extra level of indirection

File: src/lib/eap_aka_sim/state_machine.c [hunk 16]
```diff
@@ -3080,7 +3080,7 @@ static rlm_rcode_t sim_start_response_process(module_ctx_t const *mctx,
 	 *	if pseudonym resolution went ok.
 	 */
 	case FR_IDENTITY_TYPE_VALUE_PSEUDONYM:
-		if (sim_start_selected_version_check(request, from_peer, eap_aka_sim_session) < 0) goto failure;
+		if (sim_start_selected_version_check(request, &from_peer, eap_aka_sim_session) < 0) goto failure;
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:3083]: Extra level of indirection

File: src/lib/eap_aka_sim/state_machine.c [hunk 17]
```diff
@@ -3097,7 +3097,7 @@ static rlm_rcode_t sim_start_response_process(module_ctx_t const *mctx,
 	 *	later.
 	 */
 	case FR_IDENTITY_TYPE_VALUE_PERMANENT:
-		if (sim_start_selected_version_check(request, from_peer, eap_aka_sim_session) < 0) goto failure;
+		if (sim_start_selected_version_check(request, &from_peer, eap_aka_sim_session) < 0) goto failure;
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:3100]: Extra level of indirection

File: src/lib/eap_aka_sim/vector.c [hunk 18]
```diff
@@ -102,7 +102,7 @@ static int vector_gsm_from_ki(request_t *request, fr_pair_t *vps, int idx, fr_ak
 	 */
 	version_vp = fr_pair_find_by_da(&vps, attr_sim_algo_version);
 	if (!version_vp) {
-		if (vector_opc_from_op(request, &opc_p, opc_buff, vps, ki_vp->vp_octets) < 0) return -1;
+		if (vector_opc_from_op(request, &opc_p, opc_buff, &vps, ki_vp->vp_octets) < 0) return -1;
```
Comment (arr2036) [inline src/lib/eap_aka_sim/vector.c]: Extra level of indirection

File: src/lib/eap_aka_sim/vector.c [hunk 19]
```diff
@@ -111,7 +111,7 @@ static int vector_gsm_from_ki(request_t *request, fr_pair_t *vps, int idx, fr_ak
 	} else {
 		version = version_vp->vp_uint32;
 		if (version == FR_SIM_ALGO_VERSION_VALUE_COMP128_4) {
-			if (vector_opc_from_op(request, &opc_p, opc_buff, vps, ki_vp->vp_octets) < 0) return -1;
+			if (vector_opc_from_op(request, &opc_p, opc_buff, &vps, ki_vp->vp_octets) < 0) return -1;
```
Comment (arr2036) [inline src/lib/eap_aka_sim/vector.c]: Extra level of indirection

File: src/lib/eap_aka_sim/vector.c [hunk 20]
```diff
@@ -482,7 +482,7 @@ static int vector_umts_from_ki(request_t *request, fr_pair_t *vps, fr_aka_sim_ke
 		uint8_t 	opc_buff[MILENAGE_OPC_SIZE];
 		uint8_t	const	*opc_p;
 
-		if (vector_opc_from_op(request, &opc_p, opc_buff, vps, ki_vp->vp_octets) < 0) return -1;
+		if (vector_opc_from_op(request, &opc_p, opc_buff, &vps, ki_vp->vp_octets) < 0) return -1;
```
Comment (arr2036) [inline src/lib/eap_aka_sim/vector.c]: Extra level of indirection

File: src/lib/eap_aka_sim/vector.c [hunk 21]
```diff
@@ -337,7 +337,7 @@ int fr_aka_sim_vector_gsm_from_attrs(request_t *request, fr_pair_t *vps,
 	switch (*src) {
 	default:
 	case AKA_SIM_VECTOR_SRC_KI:
-		ret = vector_gsm_from_ki(request, vps, idx, keys);
+		ret = vector_gsm_from_ki(request, &vps, idx, keys);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/vector.c]: Extra level of indirection

File: src/lib/eap_aka_sim/vector.c [hunk 22]
```diff
@@ -347,7 +347,7 @@ int fr_aka_sim_vector_gsm_from_attrs(request_t *request, fr_pair_t *vps,
 		FALL_THROUGH;
 
 	case AKA_SIM_VECTOR_SRC_TRIPLETS:
-		ret = vector_gsm_from_triplets(request, vps, idx, keys);
+		ret = vector_gsm_from_triplets(request, &vps, idx, keys);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/vector.c]: Extra level of indirection

File: src/lib/eap_aka_sim/vector.c [hunk 23]
```diff
@@ -357,7 +357,7 @@ int fr_aka_sim_vector_gsm_from_attrs(request_t *request, fr_pair_t *vps,
 		FALL_THROUGH;
 
 	case AKA_SIM_VECTOR_SRC_QUINTUPLETS:
-		ret = vector_gsm_from_quintuplets(request, vps, idx, keys);
+		ret = vector_gsm_from_quintuplets(request, &vps, idx, keys);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/vector.c]: Extra level of indirection

File: src/lib/eap_aka_sim/vector.c [hunk 24]
```diff
@@ -771,7 +771,7 @@ int fr_aka_sim_vector_umts_from_attrs(request_t *request, fr_pair_t *vps,
 		FALL_THROUGH;
 
 	case AKA_SIM_VECTOR_SRC_QUINTUPLETS:
-		ret = vector_umts_from_quintuplets(request, vps, keys);
+		ret = vector_umts_from_quintuplets(request, &vps, keys);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/vector.c]: Extra level of indirection

**Review by arr2036 — CHANGES_REQUESTED** [review]

File: src/lib/eap_aka_sim/vector.c [hunk 25]
```diff
@@ -761,7 +761,7 @@ int fr_aka_sim_vector_umts_from_attrs(request_t *request, fr_pair_t *vps,
 	switch (*src) {
 	default:
 	case AKA_SIM_VECTOR_SRC_KI:
-		ret = vector_umts_from_ki(request, vps, keys);
+		ret = vector_umts_from_ki(request, &vps, keys);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/vector.c]: Extra level of indirection

File: src/modules/rlm_json/rlm_json.c [hunk 26]
```diff
@@ -285,7 +285,7 @@ static xlat_action_t json_encode_xlat(TALLOC_CTX *ctx, fr_cursor_t *out, request
 	 */
 	MEM(vb = fr_value_box_alloc_null(ctx));
 
-	json_str = fr_json_afrom_pair_list(vb, json_vps, format);
+	json_str = fr_json_afrom_pair_list(vb, &json_vps, format);
```
Comment (arr2036) [inline src/modules/rlm_json/rlm_json.c:288]: Double check this is correct

**Review by arr2036 — APPROVED** [review]

File: src/lib/curl/base.c [hunk 27]
```diff
@@ -235,7 +235,7 @@ int fr_curl_response_certinfo(request_t *request, fr_curl_io_request_t *randle)
 			/*
 			 *	Print out all the pairs we have so far
 			 */
-			log_request_pair_list(L_DBG_LVL_2, request, cert_vps, NULL);
+			log_request_pair_list(L_DBG_LVL_2, request, &cert_vps, NULL);
```
Comment (arr2036) [inline src/lib/curl/base.c:238]: The actual stack variable needs to change to a `fr_pair_list_t` as well

File: src/bin/unit_test_attribute.c [hunk 28]
```diff
@@ -1796,7 +1796,7 @@ static size_t command_encode_proto(command_result_t *result, command_file_ctx_t
 		RETURN_OK_WITH_ERROR();
 	}
 
-	slen = tp->func(cc->tmp_ctx, head, cc->buffer_start, cc->buffer_end - cc->buffer_start, encoder_ctx);
+	slen = tp->func(cc->tmp_ctx, &head, cc->buffer_start, cc->buffer_end - cc->buffer_start, encoder_ctx);
```
Comment (arr2036) [inline src/bin/unit_test_attribute.c:1799]: Stack variable needs to change to `fr_pair_list_t`

File: src/lib/eap_aka_sim/state_machine.c [hunk 29]
```diff
@@ -874,13 +874,13 @@ static int common_encode(request_t *request, eap_session_t *eap_session, uint16_
 
 
 	RDEBUG2("Encoding attributes");
-	log_request_pair_list(L_DBG_LVL_2, request, head, NULL);
+	log_request_pair_list(L_DBG_LVL_2, request, &head, NULL);
```
Comment (arr2036) [inline src/lib/eap_aka_sim/state_machine.c:877]: Stack variable needs to be an `fr_pair_list_t` too

File: src/lib/redis/cluster.c [hunk 30]
```diff
@@ -337,7 +337,7 @@ static void _cluster_node_conf_apply(fr_pool_t *pool, void *opaque)
 		if (node->cluster->trigger_args) MEM(fr_pair_list_copy(node->cluster, &args,
 								      &node->cluster->trigger_args) >= 0);
 
-		fr_pool_enable_triggers(pool, node->cluster->trigger_prefix, args);
+		fr_pool_enable_triggers(pool, node->cluster->trigger_prefix, &args);
```
Comment (arr2036) [inline src/lib/redis/cluster.c:340]: Args should be `fr_pair_list_t args;`

File: src/lib/server/exfile.c [hunk 31]
```diff
@@ -91,7 +91,7 @@ static inline void exfile_trigger_exec(exfile_t *ef, request_t *request, exfile_
 	fr_cursor_prepend(&cursor, vp);
 
 	snprintf(name, sizeof(name), "%s.%s", ef->trigger_prefix, name_suffix);
-	trigger_exec(request, ef->conf, name, false, args);
+	trigger_exec(request, ef->conf, name, false, &args);
```
Comment (arr2036) [inline src/lib/server/exfile.c:94]: Args should be `fr_pair_list_t args`

File: src/lib/server/paircmp.c [hunk 32]
```diff
@@ -215,7 +215,7 @@ static int generic_cmp(UNUSED void *instance,
 
 		if (xlat_eval(value, sizeof(value), request, name, NULL, NULL) < 0) return 0;
 
-		MEM(vp = fr_pair_afrom_da(req, check->da));
+		MEM(vp = fr_pair_afrom_da(*request_list, check->da));
```
Comment (arr2036) [inline src/lib/server/paircmp.c:218]: this isn't right... not sure the original code was right either.  I think this should be `request->packet`

File: src/lib/server/paircmp.c [hunk 33]
```diff
@@ -531,7 +531,7 @@ static int paircmp_func(request_t *request,
 
 	if (!request) return -1; /* doesn't exist, don't compare it */
 
-	return paircmp_pairs(request, check, request_list);
+	return paircmp_pairs(request, check, *request_list);
```
Comment (arr2036) [inline src/lib/server/paircmp.c:534]: This doesn't seem right.  Seems like paircmp_pairs probably needs updating too?

File: src/lib/server/paircmp.c [hunk 34]
```diff
@@ -644,7 +644,7 @@ int paircmp(request_t *request,
 		/*
 		 *	OK it is present now compare them.
 		 */
-		compare = paircmp_func(request, auth_item, check_item, check);
+		compare = paircmp_func(request, &auth_item, check_item, &check);
```
Comment (arr2036) [inline src/lib/server/paircmp.c:647]: Are these lists or single items?

File: src/lib/server/pool.c [hunk 35]
```diff
@@ -256,7 +256,7 @@ static inline void fr_pool_trigger_exec(fr_pool_t *pool, request_t *request, cha
 	if (!pool->triggers_enabled) return;
 
 	snprintf(name, sizeof(name), "%s.%s", pool->trigger_prefix, event);
-	trigger_exec(request, pool->cs, name, true, pool->trigger_args);
+	trigger_exec(request, pool->cs, name, true, &pool->trigger_args);
```
Comment (arr2036) [inline src/lib/server/pool.c:259]: trigger_args probably needs to be a `fr_pair_list_t`

File: src/lib/server/tmpl_eval.c [hunk 36]
```diff
@@ -1400,7 +1400,7 @@ int tmpl_copy_pair_children(TALLOC_CTX *ctx, fr_pair_t **out, request_t *request
  *	- -2 if list could not be found (doesn't exist in current #request_t).
  *	- -3 if context could not be found (no parent #request_t available).
  */
-int tmpl_find_vp(fr_pair_t **out, request_t *request, tmpl_t const *vpt)
+int tmpl_find_vp(fr_pair_list_t *out, request_t *request, tmpl_t const *vpt)
```
Comment (arr2036) [inline src/lib/server/tmpl_eval.c:1403]: This one's a single vp

File: src/lib/server/tmpl_eval.c [hunk 37]
```diff
@@ -1429,7 +1429,7 @@ int tmpl_find_vp(fr_pair_t **out, request_t *request, tmpl_t const *vpt)
  *	- -2 if list could not be found (doesn't exist in current #request_t).
  *	- -3 if context could not be found (no parent #request_t available).
  */
-int tmpl_find_or_add_vp(fr_pair_t **out, request_t *request, tmpl_t const *vpt)
+int tmpl_find_or_add_vp(fr_pair_list_t *out, request_t *request, tmpl_t const *vpt)
```
Comment (arr2036) [inline src/lib/server/tmpl_eval.c:1432]: Single vp

File: src/lib/unlang/module.c [hunk 38]
```diff
@@ -534,7 +534,7 @@ rlm_rcode_t unlang_module_yield_to_tmpl(TALLOC_CTX *ctx, fr_value_box_t **out, i
 	/*
 	 *	Push the xlat function
 	 */
-	if (unlang_tmpl_push(ctx, out, request, vpt, vps, status) < 0) return RLM_MODULE_FAIL;
+	if (unlang_tmpl_push(ctx, out, request, vpt, *vps, status) < 0) return RLM_MODULE_FAIL;
```
Comment (arr2036) [inline src/lib/unlang/module.c:537]: unlang_tmpl_push probably needs to take an `fr_pair_list_t`

File: src/lib/tls/validate.c [hunk 39]
```diff
@@ -159,7 +159,7 @@ int fr_tls_validate_cert_cb(int ok, X509_STORE_CTX *x509_ctx)
 			/*
 			 *	Print out all the pairs we have so far
 			 */
-			log_request_pair_list(L_DBG_LVL_2, request, cert_vps, "&session-state.");
+			log_request_pair_list(L_DBG_LVL_2, request, &cert_vps, "&session-state.");
```
Comment (arr2036) [inline src/lib/tls/validate.c:162]: cert_vps probably wants to be a `fr_pair_list_t` on the stack.

File: src/lib/server/paircmp.c (hunk 32 above)
Comment (ndptech) [reply src/lib/server/paircmp.c:218]: I attempted to make the function parameters more consistent in the different comparison functions - so in this instance req was renamed to request_list.  So my change here retains existing functionality.

**Pull request merged**